File: two_pointers/naive/rotate_list.py

from typing import Optional

class ListNode:
  def __init__(self, val: int = 0, next: Optional['ListNode'] = None):
    self.val = val
    self.next = next


def rotateRightNaive(head: Optional[ListNode], k: int) -> Optional[ListNode]:
  if not head or not head.next or k == 0:
    return head

  # Count the number of nodes
  n = 1
  tail = head
  while tail.next:
    tail = tail.next
    n += 1

  # Adjust k to avoid unnecessary rotations
  k %= n
  if k == 0:
    return head

  # Perform k rotations
  for _ in range(k):
    prev = None
    curr = head
    while curr.next:
      prev = curr
      curr = curr.next
    # Rotate the list
    curr.next = head
    head = curr
    prev.next = None

  return head
  
print(rotateRightNaive([1,2,3,4,5],2))

File: two_pointers/naive/rotateList.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func rotateRightNaive(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil || k == 0 {
		return head
	}

	// Count the number of nodes
	n := 1
	tail := head
	for tail.Next != nil {
		tail = tail.Next
		n++
	}

	// Adjust k to avoid unnecessary rotations
	k %= n
	if k == 0 {
		return head
	}

	// Perform k rotations
	for i := 0; i < k; i++ {
		var prev *ListNode
		curr := head

		for curr.Next != nil {
			prev = curr
			curr = curr.Next
		}

		// Rotate the list
		curr.Next = head
		head = curr
		prev.Next = nil
	}

	return head
}
